### ISSUE TITLE
Dilated conv in v1.0.0 is too slow.

### ISSUE THREAD

**Issue by qinjian623** [pytorch/pytorch#15054]
## 🐛 Bug
Hi, i've ported my model to 1.0.0,  then found out my model much slower then it on v0.4.1.
So I did some tests to locate the bottleneck.

It seems dilated conv in v1.0.0 is too slow.

## To Reproduce

code sample as below:
```python
import torch
import time
import torch.nn as nn

d = 3
m = nn.Conv2d(256, 256, 3, 1, padding=d, dilation=d, bias=False)
m.cuda()
total = 0.0

for _ in range(100):
    i = torch.rand(1, 256, 80, 45).cuda()
    s = time.time()
    r = m(i)
    torch.cuda.synchronize()
    e = time.time()
    total += (e-s)
print(torch.version.__version__)
print(total/100)
```
Result:

|version|d=1|d=2|d=3|
|---|---|---|---|
|0.4.1|0.00681|0.0014|0.0014|
|1.0.0|0.00328|0.0131|0.0131|

## Expected behavior
Running time should be close.

## Environment
Collecting environment information...
PyTorch version: 1.0.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 18.04.1 LTS
GCC version: (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
CMake version: version 3.12.2

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti

Nvidia driver version: 410.79
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_static_v7.a

Versions of relevant libraries:
[pip3] numpy (1.15.4)
[pip3] torch (1.0.0)
[pip3] torch-nightly (1.0.0.dev20181207)
[pip3] torchvision (0.2.1)
[pip3] torchvision-nightly (0.2.1)
[conda] blas                      1.0                         mkl  
[conda] cuda100                   1.0                           0    pytorch
[conda] magma-cuda100             2.4.0                         1    pytorch
[conda] mkl                       2018.0.3                      1  
[conda] mkl-include               2019.1                      144  
[conda] mkl-service               1.1.2            py37h90e4bf4_5  
[conda] mkl_fft                   1.0.6            py37h7dd41cf_0  
[conda] mkl_random                1.0.1            py37h4414c95_1  
[conda] pytorch                   1.0.0           py3.7_cuda10.0.130_cudnn7.4.1_1  [cuda100]  pytorch
[conda] torch                     1.0.0a0+48a361c           <pip>
[conda] torchvision               0.2.1                      py_2    pytorch

## Additional context
No.

**Comment by albanD** [comment]
Hi,

Could you add another `torch.cuda.synchronize()` before the first time.time() to avoid measuring the time to copy the Tensor please?
Also I assume you use cudnn? Could you check if using `torch.backends.cudnn.benchmark=True` gives different results please?

**Comment by qinjian623** [comment]
Hi, 

cudnn.benchmark=True and synchronize before first time:

|v|d1|d2|d3|
|---|---|---|---|
|0.4.1|0.00160|0.00111|0.00112|
|1.0.0|0.00168|0.00121|0.00129|

Now they're close enough.

Thank you, @albanD 


Should I close this issue?

**Comment by albanD** [comment]
We may want to change the way the default algorithm is picked for cudnn to avoid such inconsistencies but that's a different issue altogether.

**Comment by zimenglan-sysu-512** [comment]
hi @soumith 
without setting `torch.backends.cudnn.benchmark` explictly, the speed is twice or triple slower. after setting it to True, 10~12 time slower. it is wired. 
i use [maskrcnn-benchmark](https://github.com/facebookresearch/maskrcnn-benchmark) instead.

hi @knazeri 
do u solve it?

**Comment by zimenglan-sysu-512** [comment]
hi @knazeri 
have u ever tried muilt-scales traing?

**Comment by soumith** [comment]
This problem comes from CuDNN.

CuDNN has multiple algorithms to do convolutions, that they can choose from. They have a heuristic function that picks a particular algorithm on a particular GPU, for specific sizes of convolution / dilation etc.

In CuDNN 7.4, which is the latest version, and which is what PyTorch ships with, it looks like they introduced performance regression in this version compared to CuDNN 7.1 (which PyTorch used before).

`torch.backends.cudnn.benchmark=True` tells PyTorch / CuDNN to try every algorithm for a given {input size, kernel size, kernel stride, dilation, padding}, and pick the fastest.

The reason it fixes the problem for @knazeri is because they must be doing something like image classification, where the `input size` does not change across iterations. So, CuDNN only searches all algorithms (for the fastest algorithm) once in training -- when it encounters `input size` for the first time.

The reason it makes @zimenglan-sysu-512 's problem much worse is because they must be doing object detection or something similar, where images of different `input size` are passed in at every iteration, so CuDNN is searching all algorithms every iteration (because `input size` changes at every iteration), and searching all the list of algorithms takes 10x longer.

**Comment by soumith** [comment]
cc: @ngimel @csarofeen can we report this regression to the CuDNN team / help fix it for next release?

**Comment by syed-ahmed** [comment]
@soumith I'm working on this and will be pushing the patch this week! Tracking it here: https://github.com/pytorch/pytorch/pull/15706

**Comment by MisonoMayu** [comment]
I am expirencing the same issue with dynamic batch and dilated conv on my 1070ti(2080ti and 2080 seems not affected) However I found that set  torch.backends.cudnn.deterministic=True could be much better but still slightly slower than it was in 0.4.1.

**Comment by zzd1992** [comment]
I am experiencing the same issue. Worse, set cudnn.deterministic=True can't improve the speed on my Tian1080.

**Comment by Multihuntr** [comment]
Just wanted to share a little benchmarking I did, for those that follow.

```python
import time
import random
import torch
import torch.nn as nn
# uncomment this to gain performance:
# torch.backends.cudnn.deterministic=True

dilation = 4
conv = nn.Conv2d(47, 64, 3, padding=dilation, dilation=dilation, stride=1).cuda()
def mk_input():
  return torch.rand((random.randint(2, 15), 47, 24, 40), device='cuda:0')

for x in range(10):
  outp = conv(mk_input())

a = time.perf_counter()
for x in range(1000):
  outp = conv(mk_input())
b = time.perf_counter()
print(b-a)
# gives 0.475 with deterministic=True, and 
# 1.615 with deterministic=False
```
Using pytorch `v1.0.1` in a docker image, in docker using something approximating:
```Dockerfile
FROM nvidia/cuda:10.0-devel-ubuntu16.04
# ... python 3.6.5 installed, actual commands not printed for brevity...
RUN pip install http://download.pytorch.org/whl/cu100/torch-1.0.1-cp36-cp36m-linux_x86_64.whl
```

**Comment by dbbert** [comment]
> cc: @ngimel @csarofeen can we report this regression to the CuDNN team / help fix it for next release?

Do you know if this issue is fixed in the latest CuDNN release (7.5.0)? When will pytorch start using this release?

I'm experiencing the same ~3x slowdown with a dilation-heavy model compared to pytorch 0.4, and ```torch.backends.cudnn.benchmark=True``` doesn't improve things. I'm training with a fixed input size.

**Comment by ngimel** [comment]
@dbbert the regressions reported in this issue so far were the failure of cudnn heuristics, and to the best of my knowledge they have not been fixed in cudnn 7.5, however, they should have been fixed by cudnn.benchmark=True. If cudnn.benchmark=True does not fix the issue for you, can you please provide a small repro that we can look at?

**Comment by dbbert** [comment]
The slowdown of my model was caused by something else, see the edit in my previous message. Sorry for the confusion.

**Comment by DHZS** [comment]
I found a workaround for multi-scale training. Just use `torch.backends.cudnn.benchmark = True` and `torch.backends.cudnn.benchmark = False` to wrap your dilated conv operation. Like this:
```
import torch
import torch.nn as nn

...

conv = nn.Conv2d(256, 256, 3, 1, padding=3, dilation=3)

...

torch.backends.cudnn.benchmark = True
x = conv(x)
torch.backends.cudnn.benchmark = False

...
```

**Comment by soumith** [comment]
Hey everyone,

This issue might be fixed by cudnn 7.5.1, which was shipped in pytorch 1.1.
Please give it a try

**Comment by Multihuntr** [comment]
:heart: :heart: :heart: 

Using my previous test script. I created 2 Dockerfiles, the only difference being which version of pytorch it uses.

Here are the results:

```
                    |    Pytorch    |    Pytorch    
                    |      1.0      |      1.1      
--------------------|---------------|---------------
deterministic=False |    1.0616     |    0.1396     
--------------------|---------------|---------------
deterministic=True  |    0.2806     |    0.2948      
```

**Comment by tcapelle** [comment]
@Multihuntr  Using your script I get a very poor performance on my brand new RTX 2070 Super.
